Tweak switcher to not scale icons down

jasmussen opened the pull request:
This PR fixes so the new block icons are not scaled down, and therefore blurry, in the switcher.

Additionally it removes the gray background color from the "block chip". This is in response to similar changes elswhere in the UI.

It also adds a little dropdown arrow on hover, but not sure that adds much value, in use just this one place.

GIF:

![switcher](https://user-images.githubusercontent.com/1204802/44581642-d5e49980-a79e-11e8-9e9d-b5eb275e4f4d.gif)

Commit:
Tweak switcher to not scale icons down

This PR fixes so the new block icons are not scaled down, and therefore blurry, in the switcher.

Additionally it removes the gray background color from the "block chip". This is in response to similar changes elswhere in the UI.

It also adds a little dropdown arrow on hover, but not sure that adds much value.

mtias commented:
I wonder if we should do a more heavy-handed approach when *hovering* the block switcher. Like this, in conjunction with the tooltip:

![image 7](https://user-images.githubusercontent.com/548849/44581801-82268000-a79f-11e8-9df6-c0f2c33b33d3.png)

afercia commented:
> It also adds a little dropdown arrow on hover, but not sure that adds much value, in use just this one place.

No a11y objections but I'd say it should appear also on focus (or maybe always be displayed?). Just checked and there's no arrow on focus. Notice also the blue outline focus style, not sure that's desired:

<img width="790" alt="screen shot 2018-08-24 at 13 51 06" src="https://user-images.githubusercontent.com/1682452/44583320-10513500-a7a5-11e8-9875-c82086ccd757.png">

jasmussen commented:
So, how about this icon?

![screen shot 2018-08-24 at 14 07 51](https://user-images.githubusercontent.com/1204802/44583794-1c3df680-a7a7-11e8-87df-c6388564a992.png)

Here's the SVG in case you'd like to commandeer this PR, I have to step away until monday:

```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M6.5 8.9c.6-.6 1.4-.9 2.2-.9h6.9l-1.3 1.3 1.4 1.4L19.4 7l-3.7-3.7-1.4 1.4L15.6 6H8.7c-1.4 0-2.6.5-3.6 1.5l-2.8 2.8 1.4 1.4 2.8-2.8zm13.8 2.4l-2.8 2.8c-.6.6-1.3.9-2.1.9h-7l1.3-1.3-1.4-1.4L4.6 16l3.7 3.7 1.4-1.4L8.4 17h6.9c1.3 0 2.6-.5 3.5-1.5l2.8-2.8-1.3-1.4z"/></svg>
```

mtias commented:
Let's try it! And definitely also replace on focus.

ZebulanStanphill commented:
Quoting a comment by @rosswintle in #9127:
>I note that I was working on an iPad when I was finding it difficult. This should be an interface that does not depend on hover. It should be discoverable on touch-only devices too!

ZebulanStanphill commented:
Wrong issue or not, I think it is probably still a good point. Just showing the switcher icon on hover/focus does not cover touch-screen users. Perhaps the switcher icon should always be shown in the corner or on the side of the button?

rosswintle commented:
Ignore me. Have been on holiday and seem to have forgotten how to use GitHub. 🤦‍♂️

karmatosed commented:
> Perhaps the switcher icon should always be shown in the corner or on the side of the button?

We could always have a different solution for mobile.

Commit:
Add hover icon to switcher.

jasmussen commented:
Pushed a few fixes. On desktop breakpoints, you'll see this:

![desktop](https://user-images.githubusercontent.com/1204802/44655640-2efa3a80-a9f6-11e8-99f7-d7ae3b3c42cb.gif)

On mobile breakpoints, you'll see this:

![mobile](https://user-images.githubusercontent.com/1204802/44655661-3faab080-a9f6-11e8-9d08-9d7ae8ad9603.gif)

Your thoughts are welcome.

Commit:
Polish, make color friendly

jasmussen commented:
I pushed a few fixes to improve mobile, add consistency, and refine the code. I also made sure it was friendly to colored icons. Though note that it appears the inline styles for foreground and background colors are not output as part of the `hasColors` prop, CC: @jorgefilipecosta.

Here's how it looks now:

![transform](https://user-images.githubusercontent.com/1204802/44659732-7b00ab80-aa05-11e8-9f10-fd3abbb09b34.gif)

One thing that seems very positive about the idea of having a fixed width switcher, is that *it anchors the block*. As you transform from one block to another and everything changes, one thing stays both in place, and due to the transform icon showing on focus, stays the same as well. This anchoring has a real emotional effect and makes it feel resilient, to me.

ZebulanStanphill commented:
Wow, that feels _very_ nice! I think this is definitely an improvement over `master`.

jasmussen commented:
> This would eliminate having to hover over that icon to see the convert block type option. As it stands, the icon mimicking the block being edited acts like a visual label, something that already exists in the top right corner of the block.

We've been trying to balance things for a while here, with the hover outlines and hover label and the switcher. It's a very very delicate balance to achieve our goals and keep everyone happy. In this case, the Switcher interface is not available for every block, only blocks where transformations are reversible (i.e. lossless). If it is to double as a block type indicator, which I think is a totally fair proposal, we need to be mindful that it doesn't cause confusion with its dual roles. But I'll give this a shot.

Note that as a separate exploration, we might show the block type icon for the selected block in the top toolbar as part of explorations into nested block UIs. That would essentially always show the icon of the selected block, sidebar open or not.

TheJeffr0 commented:
Ahh ok, now I know why some blocks had the convert block option and some didn't. I thought any block could be converted to any other block. Knowing that now, it doesn't make sense to show the block change icon all the time.

jasmussen commented:
> Has there been any consideration given to always showing the transform icon instead of the block icon? 

Yes, I think that's a fair option too. The way it's designed now is to actually be a callback to the classic paragraph dropdown, but just different enough to imply that it has augmented functionality. 

Given the attention this issue is receiving currently, how would you feel about us getting this merged in to the next release, getting a feel for it, and then revisiting if need be?

hedgefield commented:
I'd say it's definitely more helpful than the current implementation, shipping and iterating seems fine.

tofumatt reviewed (changes requested):
This is cool; I'm fine with the code and the implantation but it's quite broken in IE11:

![2018-08-28 12 11 58](https://user-images.githubusercontent.com/90871/44719611-c4133700-aabb-11e8-8af4-ea8179e1146d.gif)

I can try to help with a fix a bit later, but I'm a bit pressed for time this second 😅

karmatosed commented on packages/editor/src/components/block-switcher/style.scss:
```diff
@@ -19,28 +22,53 @@
-			height: 21px;
-			width: 21px;
-		}
+	// Paint a dropdown.
```
🎨

tofumatt replied:
`bob-ross.gif`

Commit:
Address feedback and fix for IE11.

Commit:
Additional IE11 fix.

jasmussen commented:
Thanks for the review @tofumatt, I believe I have addressed all points. Comments were clarified, SVG attributes added (btw let's also look at https://github.com/webpack-contrib/svg-inline-loader), and IE11 should be addressed:

![ie11](https://user-images.githubusercontent.com/1204802/44772369-f0878b80-ab6d-11e8-88a1-c252bdc10c01.gif)

mtias reviewed (approved):
Love how this turned out.

tofumatt commented:
Tested locally and works great. I am wary about merging with failing tests, even when they are Docker-related. 🤷‍♂️

jasmussen committed:
Make the dropdown arrow slightly larger.

jasmussen commented:
Made the dropdown arrow slightly larger:

![screen shot 2018-08-29 at 14 12 24](https://user-images.githubusercontent.com/1204802/44786938-fee99d80-ab95-11e8-851a-43a0458ff848.png)

Let's either wait for the tests to pass or if 3.7 needs to ship before that, yolo. Tests pass for me too.

chrisvanpatten commented:
@jasmussen I really like this and think it should land as-is!

But… would love if a slight delay on the hover animation was considered in a follow-up PR, or if the animation was something a little tamer (maybe with a slight fade-out as the block icon moves up, and a fade-in as the switcher icon moves into place?). I like the effect a lot but it feels slightly aggressive. Which is something I never thought I'd say about a small animation like this, but there you go :)

jasmussen commented:
Thanks Chris. It's intentionally fast so as to never make you feel like you need to "wait". I imagine a more elaborate animation is nice the first few times but becomes distracting after a while.

However I might be wrong, and I would love to review a tweaked animation!

chrisvanpatten commented:
@jasmussen I did a small iteration that fades the icons in and out as they move, and changes the movement speed to 0.2s.

https://www.dropbox.com/s/q18q0cug1x8it8s/demo.mov?dl=0

(Dropbox instead of GIF because a GIF was compressing too much for the subtle opacity change to be clear.)

Gist for the changed `block-switcher/style.scss` file: https://gist.github.com/chrisvanpatten/12a49e0ffd46e33c776b7f444e07c733 (changes on lines 35 and 62)

kevinwhoffman commented:
Reverting to the pilcrow (¶) icon and the addition of the dropdown arrow improve the clarity and affordance of the control. It's more obvious now that the current block is a paragraph and that it has the potential to change as indicated by the arrow.

However I'm not in favor of the hover effect with the circular arrows icon for the following reasons:

- The sliding animation and icon change is jarring and inconsistent with any of the other existing controls in the editor.
- Once a new block is selected, the arrows icon stays in place so the new block type is not reflected in the control until focus is removed.
- The icon seems to be based on the concept of "block-switching" which is an unfamiliar concept to the average user who is more likely using the control because they want to format some text.
- The icon as proposed in https://github.com/WordPress/gutenberg/pull/9310#issuecomment-415740019 looks very similar to the existing icon for reusable blocks.

Reusable block icon seen below:
![image](https://user-images.githubusercontent.com/1938671/44815919-d15d1e00-abae-11e8-9819-6c59b283b232.png)

chrisvanpatten commented:
> The icon seems to be based on the concept of "block-switching" which is an unfamiliar concept to the average user who is more likely using the control because they want to format some text.

An unfamiliar concept is not necessarily a bad concept :) There's definitely a past legacy to account for, but new concepts are inevitable with Gutenberg.

I think this PR solves the discoverability problem — at least partially — and once someone discovers the interface, its purpose is clear.

lkraav commented:
It feels weird for the "Transform" icon to stay that, after you've made the target selection. It's like the brain is looking for a confirmation "did it do the right thing?". Right now, you can only get confirmation if you mouse away from the control and the control icon circles back to the chosen target block icon.

lkraav commented:
"Transform To" should probably sentence case?
